core/rawdb/accessors_l1_message.go

```diff
@@ -208,10 +208,64 @@ func (it *L1MessageIterator) Error() error {
 	return it.inner.Error()
 }
 
-// ReadL1MessagesFrom retrieves up to `maxCount` L1 messages starting at `startIndex`.
-func ReadL1MessagesFrom(db ethdb.Database, startIndex, maxCount uint64) []types.L1MessageTx {
+// L1MessageV1Iterator is a wrapper around L1MessageIterator that allows us to iterate over L1 messages V1.
+type L1MessageV1Iterator struct {
+	db ethdb.Database
+	L1MessageIterator
+}
+
+func IterateL1MessagesV1From(db ethdb.Database, fromQueueIndex uint64) L1MessageV1Iterator {
+	return L1MessageV1Iterator{
+		db:                db,
+		L1MessageIterator: iterateL1MessagesFrom(db, fromQueueIndex),
+	}
+}
+
+func (it *L1MessageV1Iterator) Next() bool {
+	for it.L1MessageIterator.Next() {
+		// L1MessageV2StartIndex is the first queue index of L1 messages that are from L1MessageQueueV2.
+		// Therefore, we stop reading L1 messages V1 when we reach this index.
+		// We need to check in every iteration as the start index can be set in the meantime when we are reading L1 messages.
+		v2StartIndex := ReadL1MessageV2StartIndex(it.db)
+		if v2StartIndex != nil && it.QueueIndex() >= *v2StartIndex {
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+// L1MessageV2Iterator is a wrapper around L1MessageIterator that allows us to iterate over L1 messages V2.
+type L1MessageV2Iterator struct {
+	v2StartIndex *uint64
+	L1MessageIterator
+}
+
+func IterateL1MessagesV2From(db ethdb.Database, fromQueueIndex uint64) L1MessageV2Iterator {
+	v2StartIndex := ReadL1MessageV2StartIndex(db)
+
+	return L1MessageV2Iterator{
+		v2StartIndex:      v2StartIndex,
+		L1MessageIterator: iterateL1MessagesFrom(db, fromQueueIndex),
+	}
+}
+
+func (it *L1MessageV2Iterator) Next() bool {
+	if it.v2StartIndex == nil {
+		return false
+	}
+
+	for it.L1MessageIterator.Next() {
+		return it.QueueIndex() >= *it.v2StartIndex
+	}
```

Comment:
This way even after a single iteration the caller will stop consuming the iterator, is this intended?

Reply:
That's not intended. But I'm not sure how this would be the case?

Reply:
I guest my point was that using a `for` loop is confusing, since it's only possible to have 1 or 0 iterations. Easier to understand with `if`, or simply `return it.L1MessageIterator.Next() && it.QueueIndex() >= *it.v2StartIndex`

Reply:
done in 62617efda696e4c29441a4a02d15f61b9fcf5ef5